File: codeforces/1800~1899/1876/a.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0, p = 0;
  cin >> n >> p;
  vector<int> a(n);
  for (int i = 0; i < n; ++i) cin >> a[i];
  vector<int> b(n);
  for (int i = 0; i < n; ++i) cin >> b[i];

  vector<pair<int,int>> residents;
  for (int i = 0; i < n; ++i) {
    residents.emplace_back(b[i], a[i]);
  }
  sort(residents.begin(), residents.end());
  priority_queue<pair<int,int>, vector<pair<int,int>>, greater<>> pq;
  pq.push({p, n});

  i64 ans = 0;
  for (int i = 0; i < n; ++i) {
    auto [cost, left] = pq.top();
    pq.pop();

    ans += cost;

    if (--left > 0) {
      pq.push({cost, left});
    }
    pq.push(residents[i]);
  }

  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1800~1899/1876/a.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::BinaryHeap;
use std::cmp::Reverse;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let inputs = read_vec::<i64>();
    let n = inputs[0] as usize;
    let p = inputs[1];
    let a = read_vec::<i64>();
    let b = read_vec::<i64>();

    let mut pq = BinaryHeap::new();
    pq.push(Reverse((p, n as i64)));
    let mut residents = Vec::new();
    for i in 0..n {
        residents.push((b[i], a[i]));
    }
    residents.sort();

    let mut ans: i64 = 0;
    for i in 0..n {
        if let Some(Reverse((bi, ai))) = pq.pop() {
            ans += bi;
            if ai-1 > 0 {
                pq.push(Reverse((bi, ai-1)));
            }
            pq.push(Reverse(residents[i]));
        }
    }

    writeln!(writer, "{}", ans).ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
